File: python/stackQueue/nextSmaller.py
def nextSmallerElement(arr, n):
    # Write your code here.
    stack = list()
    ans = [-1] * n

    for i in range(n - 1, -1, -1):
        while len(stack) and stack[-1] >= arr[i]:
            stack.pop()

        if len(stack) and stack[-1] < arr[i]:
            ans[i] = stack[-1]

        stack.append(arr[i])

    return ans


arr = [3, 3, 1, 1]
n = len(arr)

print(nextSmallerElement(arr, n))

File: cpp/stackQueue/nextSmaller.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> nextSmaller(vector<int> nums) {
  int n = nums.size();
  vector<int> ans(n, -1);
  stack<int> st;

  for (int i = n - 1; i >= 0; i--) {
    while (!st.empty() and st.top() >= nums[i]) {
      st.pop();
    }

    if (!st.empty()) {
      ans[i] = st.top();
    }
    st.push(nums[i]);
  }
  return ans;
}

int main() {
  vector<int> nums = {1, 3, 2, 4};
  vector<int> ans = nextSmaller(nums);
  for (auto ele : ans)
    cout << ele << " ";
  cout << endl;

  return 0;
}
